RevenueCatUI/CustomerCenter/Data/PurchaseInformation.swift

```diff
@@ -46,60 +44,46 @@ struct PurchaseInformation {
-        self.price = price
-        self.expirationOrRenewal = expirationOrRenewal
-        self.productIdentifier = productIdentifier
-        self.store = store
-    }
-
-    init(entitlement: EntitlementInfo,
+    init(entitlement: EntitlementInfo? = nil,
          subscribedProduct: StoreProduct? = nil,
+         transaction: Transaction,
          dateFormatter: DateFormatter = DateFormatter()) {
         dateFormatter.dateStyle = .medium
 
+        // Title and duration from product if available
         self.title = subscribedProduct?.localizedTitle
-        self.explanation = entitlement.explanation
         self.durationTitle = subscribedProduct?.subscriptionPeriod?.durationTitle
-        self.price = entitlement.priceBestEffort(product: subscribedProduct)
-        self.expirationOrRenewal = entitlement.expirationOrRenewal(dateFormatter: dateFormatter)
-        self.productIdentifier = entitlement.productIdentifier
-        self.store = entitlement.store
-    }
 
-    init(product: StoreProduct,
-         expirationDate: Date?,
-         dateFormatter: DateFormatter = DateFormatter()) {
-        // We don't have enough information to determine if the subscription will renew or not because we
-        // are loading the information from the product without entitlement information.
-        // We also assume that the subscription is active.
-        // We also assume that the subscription will renew the earliest possible renewal date and this is the
-        // product with the earliest renewal date.
-        dateFormatter.dateStyle = .medium
-
-        self.title = product.localizedTitle
-        self.explanation = .earliestRenewal
-        self.durationTitle = product.subscriptionPeriod?.durationTitle
-        self.price = .paid(product.localizedPriceString)
-        if let dateString = expirationDate.map({ dateFormatter.string(from: $0) }) {
-            let date = PurchaseInformation.ExpirationOrRenewal.Date.date(dateString)
-            self.expirationOrRenewal = PurchaseInformation.ExpirationOrRenewal(label: .expires,
-                                                                               date: date)
+        // Use entitlement data if available, otherwise derive from transaction
+        if let entitlement = entitlement {
+            self.explanation = entitlement.explanation
+            self.expirationOrRenewal = entitlement.expirationOrRenewal(dateFormatter: dateFormatter)
+            self.productIdentifier = entitlement.productIdentifier
+            self.store = entitlement.store
+            self.price = entitlement.priceBestEffort(product: subscribedProduct)
         } else {
-            self.expirationOrRenewal = nil
+            switch transaction.type {
+            case .subscription(let isActive, let willRenew, let expiresDate):
+                self.explanation = expiresDate != nil
+                    ? (isActive ? (willRenew ? .earliestRenewal : .earliestExpiration) : .expired)
```

Comment:
oh so these are just little texts that appear at the top of the view. 

subEarliestRenewal: "This is your subscription with the earliest billing date."
subEarliestExpiration: "This is your subscription with the earliest expiration date."

So the logic is that if the sub is active and it will renew, it should say "billing date" and if it won't renew it should say "expiration date"

Reply:
Oohhhh, so it's the subscription with the earliest billing date in the _future_! I read it as "earliest billing date...in the past" and was confused 😅

Makes sense!